packages/math-utils/src/memory_calculator.ts

```diff
@@ -0,0 +1,237 @@
+    evaluateDependencies,
+    // statistics dependencies
+    maxDependencies,
+    minDependencies,
+    // logical dependencies
+    andDependencies,
+    notDependencies,
+    orDependencies,
+    xorDependencies,
+    // without that dependency 'null ?? 5', won't work
+    nullishDependencies,
+});
+const limitedEvaluate = math.evaluate;
+const limitedCompile = math.compile;
+
+// Disable potentially dangerous functions
+math.import({
+    // most important (hardly any functional impact)
+    import() { throw new Error('Function import is disabled'); },
+    createUnit() { throw new Error('Function createUnit is disabled'); },
+    reviver() { throw new Error('Function reviver is disabled'); },
+
+    // extra (has functional impact)
+    evaluate() { throw new Error('Function evaluate is disabled'); },
+    parse() { throw new Error('Function parse is disabled'); },
+    simplify() { throw new Error('Function simplify is disabled'); },
+    derivative() { throw new Error('Function derivative is disabled'); },
+    resolve() { throw new Error('Function resolve is disabled'); },
+}, { override: true });
+
+/**
+ * Safely retrieves a nested property from an object using a dot-notation string path.
+ *
+ * This is custom function designed to be injected into the math expression evaluator,
+ * allowing expressions like `get(input, 'user.settings.memory', 512)`.
+ *
+ * @param obj The source object to search within.
+ * @param path A dot-separated string representing the nested path (e.g., "input.payload.size").
+ * @param defaultVal The value to return if the path is not found or the value is `null` or `undefined`.
+ * @returns The retrieved value, or `defaultVal` if the path is unreachable.
+*/
+const customGetFunc = (obj: any, path: string, defaultVal?: number) => {
+    return (path.split('.').reduce((current, key) => current?.[key], obj)) ?? defaultVal;
+};
+
+/**
+ * Rounds a number to the closest power of 2.
+ * The result is clamped to the allowed range (ACTOR_LIMITS.MIN_RUN_MEMORY_MBYTES - ACTOR_LIMITS.MAX_RUN_MEMORY_MBYTES).
+ * @param num The number to round.
+ * @returns The closest power of 2 within min/max range.
+*/
+const roundToClosestPowerOf2 = (num: number): number | undefined => {
+    if (typeof num !== 'number' || Number.isNaN(num)) {
+        throw new Error(`Failed to round number to a power of 2.`);
+    }
+
+    // Handle 0 or negative values.
+    if (num <= 0) {
+        throw new Error(`Calculated memory value must be a positive number, greater than 0, got: ${num}`);
+    }
+
+    const log2n = Math.log2(num);
+
+    const roundedLog = Math.round(log2n);
+    const result = 2 ** roundedLog;
+
+    return Math.max(ACTOR_LIMITS.MIN_RUN_MEMORY_MBYTES, Math.min(result, ACTOR_LIMITS.MAX_RUN_MEMORY_MBYTES));
+};
+
+/**
+ * Replaces `{{variable}}` placeholders in an expression string with the variable name.
+ * Enforces strict validation to allow `{{input.*}}` paths or whitelisted `{{runOptions.*}}` keys.
+ *
```

Comment:
I think this would deserve a proper paragraph of explanation for why we do this.

Otherwise somebody will get really confused by this 😄